Introduce Quarkus for the Web documentation (#39953)
quarkusio/quarkus: docs/src/main/asciidoc/qute.adoc

The solution is located in the `qute-quickstart` link:{quickstarts-tree-url}/qute-quickstart[directory].

== Hello World with Jakarta REST

If you want to use Qute in your Jakarta REST application, you need to add an extension first:
[[hello-qute-web]]
== Hello World with Qute Web

* either `quarkus-rest-qute` if you are using Quarkus REST (formerly RESTEasy Reactive):
+
If you want to use Qute in your Quarkus Web application, add the Qute Web extension first:
[source,xml,role="primary asciidoc-tabs-target-sync-cli asciidoc-tabs-target-sync-maven"]
.pom.xml
----
<dependency>
    <groupId>io.quarkus</groupId>

>     <artifactId>quarkus-rest-qute</artifactId>

Reviewer: remind me again - was this feature removed or just moved to qute-web?

>     <groupId>io.quarkiverse.qute.web</groupId>

Reviewer: having docs in our main http docs  that uses extensions not part of platform - thats a first is it not?

Author: it is part of the platform

Author: ![image](https://github.com/quarkusio/quarkus/assets/2223984/59c43767-8c54-4382-8d02-6d4cc6a5db6f)

----
implementation("io.quarkiverse.qute.web:quarkus-qute-web")
----

All files located in the `src/main/resources/templates` directory and its subdirectories are registered as templates. Templates are validated during startup and watched for changes in the development mode.

Now, let's start with a Hello World html template:

.src/main/resources/templates/pub/hello.html
[source]
----
<h1>Hello {http:param('name', 'Quarkus')}!</h1> <1>
----
<1> `{http:param('name', 'Quarkus')}` is an expression that is evaluated when the template is rendered (Quarkus is the default value).

NOTE: Templates located in the `pub` directory are served via HTTP. Automatically, no controllers needed. For example, the template src/main/resource/templates/pub/foo.html will be served from the paths /foo and /foo.html by default.

If your application is running, you can open your browser and hit:  http://localhost:8080/hello?name=Martin

For more information about Qute Web, see the https://docs.quarkiverse.io/quarkus-qute-web/dev/index.html[Qute Web guide].

[[hello-qute-rest]]

> == Hello World with Jakarta REST

Author: I don't get it? Why would it go first? In both case you are serving templates?

The qute-web basic approach is just simpler until you need something more specific with Quarkus REST

> == Hello World with Jakarta REST

Reviewer 2: Because the Jakarta REST is IMHO a more simple and a more common use case.

You are trying to push your web story very hard in a guide that has nothing to do with your Quarkus Web guide.

For me the basic use case of using Qute with REST should come first.

(Also please don't mark conversations as resolved when they aren't as it makes it very hard to track where feedback was given)

> == Hello World with Jakarta REST

Reviewer: agreeing with gsmet here that in this guide the "fuller" web app is a bit secondary and qute with rest fits and then web app comes next

> If you want to use Qute in your Jakarta REST application, you still need to add the Qute Web extension first (see <<hello-qute-web>>) and make sure you have the Quarkus REST (formerly RESTEasy Reactive) extension.

Author: Qute Web is part of the platform, this has already been discussed

> If you want to use Qute in your Jakarta REST application, you still need to add the Qute Web extension first (see <<hello-qute-web>>) and make sure you have the Quarkus REST (formerly RESTEasy Reactive) extension.

Reviewer: @ia3andy it does reads problematically here though - it would be good to somehow capture what the discussion was and be able to answer gsmet's question why it was not added to quarkus-rest.

> If you want to use Qute in your Jakarta REST application, you still need to add the Qute Web extension first (see <<hello-qute-web>>) and make sure you have the Quarkus REST (formerly RESTEasy Reactive) extension.

Author: The idea is:
- Qute Web + REST = Qute REST
- Qute Web + RESTEasy = Qute RESTEasy

This is the concept of conditional dependency to avoid having hundreds of listed extensions just for compatibly. I am not sure if it's already configured, but it should be the same for Jackson or other case like this.
- REST + Jackson = REST Jackson (no need to add it)

> If you want to use Qute in your Jakarta REST application, you still need to add the Qute Web extension first (see <<hello-qute-web>>) and make sure you have the Quarkus REST (formerly RESTEasy Reactive) extension.

Reviewer 2: Still my problem is that we are pushing something that is not supported (yet?) as the way to go whereas there is a perfectly supported alternative that is working and has always worked.
Being in the Platform does not mean being supported.

Not everyone is (unfortunately) comfortable with using Quarkiverse extensions.

And while I don't have issues with adding info about Quarkiverse extensions here, I think it's a bit unfortunate we drop some perfectly working and supported way of doing things.

Now that's @maxandersen call here. But I have the impression you're pushing this a bit too hard on everyone.

> If you want to use Qute in your Jakarta REST application, you still need to add the Qute Web extension first (see <<hello-qute-web>>) and make sure you have the Quarkus REST (formerly RESTEasy Reactive) extension.

Reviewer: @ia3andy you have to remind me again why is it that we ended up doing the conditional dependency trick here outside in quarkiverse and not in Quarkus core?

@gsmet the platform is more about what is supported than anything else we have. we got stuff in core that is most definitely less supported :) i do agree we should have this clarified since:

a) if we can make it work better by just having the qute res stuff in core we probably should and we can remove this unfortuante mention about you must use quarkiverse for jakarta rest)
b) if we can't - we should rephrase/move this around to be more of a recommendation/suggestion.

The solution is located in the `qute-quickstart` link:{quickstarts-tree-url}/qute-quickstart[directory].

[[serving-templates]]
== Serving Qute templates via http

If you want to serve your templates via http:

1. The Qute Web extension allows you to directly serve via http templates located in `src/main/resource/templates/pub/`. In that case you don't need any Java code to "plug" the template, for example, the template `src/main/resource/templates/pub/foo.html` will be served from the paths `/foo` and `/foo.html` by default.

> 2. For finer control, you can combine it with Quarkus REST or Quarkus RESTEasy to control how your template will be served. All files located in the `src/main/resources/templates` directory and its subdirectories are registered as templates and can be injected in a REST resource.

Reviewer: ```suggestion
2. For finer control, you can combine it with Quarkus REST to control how your template will be served. All files located in the `src/main/resources/templates` directory and its subdirectories are registered as templates and can be injected in a REST resource.
```

I'm assuming its just Quarkus REST you mean to refer to here?

Author: Both are possible AFAIK, should we mention only one?

[source,xml,role="primary asciidoc-tabs-target-sync-cli asciidoc-tabs-target-sync-maven"]
.pom.xml
----
<dependency>
    <groupId>io.quarkiverse.qute.web</groupId>
    <artifactId>quarkus-qute-web</artifactId>
</dependency>
----

[source,gradle,role="secondary asciidoc-tabs-target-sync-gradle"]
.build.gradle
----
implementation("io.quarkiverse.qute.web:quarkus-qute-web")
----

> NOTE: The Qute Web extension is still using the quarkiverse group-id, it is part of the Quarkus platform (so it's is included in the bom), we are working on moving it to Quarkus Core.

Reviewer: ```suggestion
NOTE: The Qute Web extension while using the quarkiverse group-id, it is still part of the Quarkus platform.
```

keeping it simple and not try making it sound as its too much moving.
